### PR TITLE
add wire-xunit tests / fix to harness & docgen - master

### LINKED ISSUE #42: Docgen not working with xunit testresults from wiremock
When running DocGen it stucks when it runs projects that use wiremock technology in theirs tests..

**To Reproduce**
1-Write a integration test using Spring Boot 2.2.7 that uses wiremock v2.26.3
2-The test should run fine and succesfully.
3-Integrate this project into a Release Manager with DocGen.
4-Run Developer Preview on Jira
5-A SocketTimeOutException exception ocurrs.
**Expected behavior**
All tests should pass and the documentation should be generated
**Troubleshooting**

From Openshift the jenkins pod agent is killed after DocGen sends the response..In DocGen log files there are no errors available.

**Screenshots**

**Affected version (please complete the following information):**
 - OpenShift: 3.11
 - OpenDevStack 2.0

**Log Output (ensure to remove any confidential information like tokens, project names, etc.**
```

java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at unirest.shaded.org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:137)
	at unirest.shaded.org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:153)
	at unirest.shaded.org.apache.http.impl.io.SessionInputBufferImpl.readLine(SessionInputBufferImpl.java:280)
	at unirest.shaded.org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:138)
	at unirest.shaded.org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:56)
	at unirest.shaded.org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:259)
	at unirest.shaded.org.apache.http.impl.DefaultBHttpClientConnection.receiveResponseHeader(DefaultBHttpClientConnection.java:163)
	at unirest.shaded.org.apache.http.impl.conn.CPoolProxy.receiveResponseHeader(CPoolProxy.java:157)
	at unirest.shaded.org.apache.http.protocol.HttpRequestExecutor.doReceiveResponse(HttpRequestExecutor.java:273)
	at unirest.shaded.org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:125)
	at unirest.shaded.org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:272)
	at unirest.shaded.org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
	at unirest.shaded.org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)
	at unirest.shaded.org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at unirest.shaded.org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	at unirest.shaded.org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at unirest.shaded.org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at unirest.shaded.org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:56)
	at kong.unirest.apache.ApacheClient.request(ApacheClient.java:125)
	at kong.unirest.BaseRequest.asString(BaseRequest.java:180)
	at sun.reflect.GeneratedMethodAccessor10641.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1213)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1022)
	at org.codehaus.groovy.runtime.callsite.PojoMetaClassSite.call(PojoMetaClassSite.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:117)
	at org.ods.service.DocGenService.createDocument(DocGenService.groovy:37)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1213)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1022)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:42)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at com.cloudbees.groovy.cps.sandbox.DefaultInvoker.methodCall(DefaultInvoker.java:20)
Caused: kong.unirest.UnirestException
	at kong.unirest.apache.ApacheClient.request(ApacheClient.java:133)
	at kong.unirest.BaseRequest.asString(BaseRequest.java:180)
	at sun.reflect.GeneratedMethodAccessor10641.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1213)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1022)
	at org.codehaus.groovy.runtime.callsite.PojoMetaClassSite.call(PojoMetaClassSite.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:117)
	at org.ods.service.DocGenService.createDocument(DocGenService.groovy:37)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1213)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1022)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:42)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at com.cloudbees.groovy.cps.sandbox.DefaultInvoker.methodCall(DefaultInvoker.java:20)
	at org.ods.usecase.DocGenUseCase.createDocument(DocGenUseCase.groovy:32)
	at org.ods.usecase.LeVADocumentUseCase.createDTR(LeVADocumentUseCase.groovy:515)
	at org.ods.scheduler.LeVADocumentScheduler.run(LeVADocumentScheduler.groovy:291)
	at org.ods.util.PipelineUtil.executeBlockAndFailBuild(PipelineUtil.groovy:110)
	at org.ods.scheduler.LeVADocumentScheduler.run(LeVADocumentScheduler.groovy:287)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2030)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2015)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2056)
	at org.ods.scheduler.LeVADocumentScheduler.run(LeVADocumentScheduler.groovy:277)
	at phaseBuild.call(phaseBuild.groovy:41)
	at org.ods.util.MROPipelineUtil.prepareExecutePhaseForRepoNamedJob(MROPipelineUtil.groovy:530)
	at org.ods.util.PipelineUtil.executeBlockAndFailBuild(PipelineUtil.groovy:110)
	at org.ods.util.MROPipelineUtil.prepareExecutePhaseForRepoNamedJob(MROPipelineUtil.groovy:450)
	at ___cps.transform___(Native Method)
	at com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:84)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:113)
Caused: java.lang.IllegalStateException: Error: Creating document of type 'DTR' for project 'xxxx' and repo 'xxxxx-aaaaaa' in phase 'Build' and stage 'POST_EXECUTE_REPO' has failed: java.net.SocketTimeoutException: Read timed out.
	at org.ods.scheduler.LeVADocumentScheduler.run(LeVADocumentScheduler.groovy:293)
	at org.ods.util.PipelineUtil.executeBlockAndFailBuild(PipelineUtil.groovy:110)
	at org.ods.scheduler.LeVADocumentScheduler.run(LeVADocumentScheduler.groovy:287)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2030)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2015)
	at com.cloudbees.groovy.cps.CpsDefaultGroovyMethods.each(CpsDefaultGroovyMethods:2056)
	at org.ods.scheduler.LeVADocumentScheduler.run(LeVADocumentScheduler.groovy:277)
	at phaseBuild.call(phaseBuild.groovy:41)
	at org.ods.util.MROPipelineUtil.prepareExecutePhaseForRepoNamedJob(MROPipelineUtil.groovy:530)
	at org.ods.util.PipelineUtil.executeBlockAndFailBuild(PipelineUtil.groovy:110)
	at org.ods.util.MROPipelineUtil.prepareExecutePhaseForRepoNamedJob(MROPipelineUtil.groovy:450)
	at ___cps.transform___(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:83)
	at org.codehaus.groovy.reflection.CachedConstructor.doConstructorInvoke(CachedConstructor.java:77)
	at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrap.callConstructor(ConstructorSite.java:84)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:60)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:235)
	at com.cloudbees.groovy.cps.sandbox.DefaultInvoker.constructorCall(DefaultInvoker.java:25)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:97)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixArg(FunctionCallBlock.java:83)
	at sun.reflect.GeneratedMethodAccessor426.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.LocalVariableBlock$LocalVariable.get(LocalVariableBlock.java:39)
	at com.cloudbees.groovy.cps.LValueBlock$GetAdapter.receive(LValueBlock.java:30)
	at com.cloudbees.groovy.cps.impl.LocalVariableBlock.evalLValue(LocalVariableBlock.java:28)
	at com.cloudbees.groovy.cps.LValueBlock$BlockImpl.eval(LValueBlock.java:55)
	at com.cloudbees.groovy.cps.LValueBlock.eval(LValueBlock.java:16)
	at com.cloudbees.groovy.cps.Next.step(Next.java:83)
	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:174)
	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:163)
	at org.codehaus.groovy.runtime.GroovyCategorySupport$ThreadCategoryInfo.use(GroovyCategorySupport.java:129)
	at org.codehaus.groovy.runtime.GroovyCategorySupport.use(GroovyCategorySupport.java:268)
	at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:163)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:18)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:51)
	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:186)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:370)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$200(CpsThreadGroup.java:93)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:282)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:270)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:66)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:131)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:59)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Finished: FAILURE
```
**Additional context**
If all tests using wiremock are excluded from build then DocGen passes succesfully and documentation is generated.

### PR DESCRIPTION
**Pull request by clemensutschig**
closes #42

### CODE REVIEW

**Comment by clemensutschig** [comment]
@jssnull  - hurray - I have a fix for this :) @metmajer - so first it looks like there was an issue with the "footer" in these super complex cases,.. and secondly - I think we found a jvm bug .. (at least on openjdk which I am running) - that the process implementation gets stuck (with waitFor()) .... see code comments ..

@jssnull  - can you please try with the branch? @alanbort - fyi

**Comment by clemensutschig** [comment]
ping?

**Comment by clemensutschig** [comment]
@alanbort - fyi .. do you need this in 3.x as well?

**Comment by jssnull** [comment]
Hi, I'll test this pull request in a ODS 3.x test environment and I'll let you know 
Best Regards

**Comment by segator** [comment]
If i'm not wrong @clemensutschig  this is not a bug, the stderr and stdout must be pulled or the proc buffer got full so the process got stucked and waitfor() will never ends.

**Comment by clemensutschig** [comment]
> If i'm not wrong @clemensutschig this is not a bug, the stderr and stdout must be pulled or the proc buffer got full so the process got stucked and waitfor() will never ends.

this was a bug - and did cleanly reproduce (at a certain size and complexity...) - please verify

File: src/main/groovy/app/DocGen.groovy [hunk 1]
```diff
@@ -189,12 +190,23 @@ ${data.metadata.header[1]}"""])
         // Execute a command in the shell
         static private def Map shell(List<String> cmd) {
             def proc = cmd.execute()
-            proc.waitFor()
+            // for some VERY complex docs - the process implementation hangs on wait() ...
+            // switching to the below - seem to work but gets a weird NPE... 
+            // java.lang.NullPointerException: Cannot invoke method call() on null object
+            // at app.DocGen$Util.shell(DocGen.groovy:193)
+            ByteArrayOutputStream bosOut = new ByteArrayOutputStream()
```
Comment (segator) [inline src/main/groovy/app/DocGen.groovy:197]: prefered to use ```NullOutputStream``` and ```NullInputStream```, using ```ByteArrayOutputStream``` is using inecesary memory and can even cause OOM

**Comment by segator** [comment]
this behaviour is documented here: https://docs.oracle.com/javase/8/docs/api/java/lang/Process.html

 _All its standard I/O (i.e. stdin, stdout, stderr) operations will be redirected to the parent process, where they can be accessed via the streams obtained using the methods getOutputStream(), getInputStream(), and getErrorStream(). The parent process uses these streams to feed input to and get output from the subprocess. Because some native platforms only provide limited buffer size for standard input and output streams, failure to promptly write the input stream or read the output stream of the subprocess_ **may cause the subprocess to block, or even deadlock**.

File: src/main/groovy/app/DocGen.groovy (hunk 1 above)
Comment (clemensutschig) [reply src/main/groovy/app/DocGen.groovy:197]: which means you can't get ANY error message, whatsoever .. we have tested this with >2000 page tests ... any your get a 300 bytes ...

**Review by segator — APPROVED** [review]

**Comment by clemensutschig** [comment]
@jssnull  - any update?

**Comment by jssnull** [comment]
@clemensutschig sorry I was very busy, I'll take a timeslot today for test it

**Comment by jssnull** [comment]
currently testing, I'll let you know ASAP

**Comment by clemensutschig** [comment]
@jssnull - and?

**Comment by jssnull** [comment]
Hi, I'm having git connection errors with my ODS test project using Docgen, I'll let you know ASAP

**Comment by jssnull** [comment]
Hi, 
I tested a Docgen image builded using this pull request with a test-results.xml that contains 7000 lines of system-output and it failed.

This is the error log:
`[ods-mro-pipeline] **** ENDED orchestration pipeline **** (took 241026 ms)
[Pipeline] }
[Pipeline] // podTemplate
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
java.lang.RuntimeException: Error: unable to create document 'DTR (v1.0)'. DocGen responded with code: '500' and message: '{"message":"Loading pages (1/6)\n[&gt;                                                           ] 0%\r[======&gt;                                                     ] 10%\r[==============================&gt;                             ] 50%\r","stacktrace":["java.lang.IllegalStateException: Loading pages (1/6)","[>                                                           ] 0%[======>                                                     ] 10%[==============================>                             ] 50%","\tat sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)","\tat sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)","\tat sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)","\tat java.lang.reflect.Constructor.newInstance(Constructor.java:423)","\tat org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:80)","\tat org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:105)","\tat ` 

Here is attached the xml file used in a .txt file
[test-results.txt](https://github.com/opendevstack/ods-document-generation-svc/files/5566489/test-results.txt)

**Comment by clemensutschig** [comment]
@jssnull - two requests
a) please post the entire stack trace
b) please check the doc gen pod - and post the logs here as well.

thanks!°

**Comment by jssnull** [comment]
After debugging with @clemensutschig  we've discovered that the docgen service is throwing an OOM error in POD's log

`Doc ID/Version: see auto-generated cover page, --header-font-size, 10, --header-spacing, 10, --footer-center, 'Page [page] of [topage]', --footer-font-size, 10, /tmp/DTR-v1.01545279969132907775/templates/DTR.html, /tmp/document6650489645235884204.pdf] has exited with code 137
[ERROR]: Loading pages (1/6)`

I'll increase process memory limit and I'll let you know

**Comment by clemensutschig** [comment]
with the lsat commit - we also return the error code .. :)

**Comment by jssnull** [comment]
@clemensutschig I tested this, increasing PODs memory and it worker
image
 
Thanks!

**Review by jssnull — APPROVED** [review]
Worked with an xunit with 7000 lines of system-out

**Comment by clemensutschig** [comment]
super cool! I'll merge it and backport it

**Pull request merged**